contracts/hooks/FullRange.sol

```diff
@@ -0,0 +1,492 @@
+
+        // deploy erc20 contract
+
+        bytes memory bytecode = type(UniswapV4ERC20).creationCode;
+        bytes32 salt = keccak256(abi.encodePacked(key.toId()));
+
+        address poolToken;
+        assembly {
+            poolToken := create2(0, add(bytecode, 32), mload(bytecode), salt)
+        }
+
+        poolToERC20[key.toId()] = poolToken;
+
+        return FullRange.beforeInitialize.selector;
+    }
+
+    function _rebalance(IPoolManager.PoolKey calldata key) internal {
+        // if we're at a new block, rebalance
+        if (block.number > blockNumber) {
+            blockNumber = block.number;
+            // retrieve all liquidity
+            uint128 hookLiquidity = poolManager.getLiquidity(key.toId(), address(this), MIN_TICK, MAX_TICK);
+
+            HookPosition storage position = poolToHookPosition[key.toId()];
+
+            position.liquidity = hookLiquidity;
+
+            IPoolManager.ModifyPositionParams memory params = IPoolManager.ModifyPositionParams({
+                tickLower: MIN_TICK,
+                tickUpper: MAX_TICK,
+                liquidityDelta: -int256(int128(hookLiquidity))
+            });
+
+            BalanceDelta balanceDelta = hookModifyPosition(key, params);
+
+            (uint160 sqrtPriceX96,,,,,) = poolManager.getSlot0(key.toId());
+
+            uint160 sqrtRatioAX96 = TickMath.getSqrtRatioAtTick(MIN_TICK);
+            uint160 sqrtRatioBX96 = TickMath.getSqrtRatioAtTick(MAX_TICK);
+
+            uint128 liquidity = LiquidityAmounts.getLiquidityForAmounts(
+                sqrtPriceX96,
+                sqrtRatioAX96,
+                sqrtRatioBX96,
+                uint256(uint128(balanceDelta.amount0())),
+                uint256(uint128(balanceDelta.amount1()))
+            );
+
+            params = IPoolManager.ModifyPositionParams({
+                tickLower: MIN_TICK,
+                tickUpper: MAX_TICK,
+                liquidityDelta: int256(int128(liquidity))
+            });
+
+            // reinvest everything
+            modifyPosition(key, params);
+
+            // update position
+
+            Position.Info memory posInfo = poolManager.getPosition(key.toId(), address(this), MIN_TICK, MAX_TICK);
+            uint256 feeGrowthInside0LastX128 = posInfo.feeGrowthInside0LastX128;
+            uint256 feeGrowthInside1LastX128 = posInfo.feeGrowthInside1LastX128;
+
+            position.feeGrowthInside0LastX128 = feeGrowthInside0LastX128;
+            position.feeGrowthInside1LastX128 = feeGrowthInside1LastX128;
+            position.tokensOwed0 = 0;
+            position.tokensOwed1 = 0;
+        }
+    }
+
+    function beforeModifyPosition(
+        address sender,
+        IPoolManager.PoolKey calldata key,
+        IPoolManager.ModifyPositionParams calldata params
+    ) external override returns (bytes4) {
+        // check msg.sender
+        require(sender == address(this), "sender must be hook");
+
+        // check full range
+        require(
```

Comment:
if we're checking that the sender is address this, isn't verifying min/max tick redundant?